cpplint.py

```diff
@@ -938,6 +940,33 @@ def unicode_escape_decode(x):
 # category should be suppressed for every line.
 _global_error_suppressions = {}
 
+class BlockErrorSuppressions(set):
+  """ Modified set that tracks which line a suppression block was started on """
+  def __init__(self, *args, **kwargs):
+    super().__init__(*args, **kwargs)
+    self._linenum = -1
+
+  def __str__(self):
+    if self.linenum < 0:
+      return 'NONE'
+    return 'Categories suppressed from line %s: %s' % \
+           (self.linenum, 'ALL' if None in self else ", ".join(c for c in self))
+
+  def clear(self):
+    super().clear()
+    self._linenum = -1
+
+  @property
+  def linenum(self):
+    return self._linenum
+
+  @linenum.setter
+  def linenum(self, linenum):
+    assert linenum > 0
+    self._linenum = linenum
+
+_block_error_suppressions = BlockErrorSuppressions()
```

Comment:
just thinking aloud right now: Did you consider unifying _block_error_suppressions and _error_suppressions? One seems like a generalization of the other.

Reply:
It could be done I'm not sure it would simplify things too much since you would then have to add code to trigger the auto clearing of the single line block on the line after it expires or track ranges of blocks which are suppressed. I can play around with this if you like.

Reply:
I added a commit which converts the current approach of separate data structs to track each type suppression into a single data structure which can be used to track any type of suppression